File: Pather/Pather/Person.ts
/// <reference path="constants.ts" />

module Pather {
    export class Person {
        x: number;
        y: number;
        squareX: number;
        squareY: number;
        speed: number;
        path: AStar.AStarPath[];
        rePathFindPosition: utils.Point;
        animations: utils.AnimationPoint[];
        game: Game;

        constructor(game: Game) {
            this.x = 0;
            this.y = 0;
            this.squareX = 0;
            this.squareY = 0;
            this.speed = 50;
            this.path = [];
            this.rePathFindPosition = undefined;
            this.game = game;
        }

        init(squareX: number, squareY: number) {
            this.squareX = squareX;
            this.squareY = squareY;
            this.x = this.squareX * Constants.squareSize;
            this.y = this.squareY * Constants.squareSize;
            this.animations = [];

        }

        draw(context: CanvasRenderingContext2D, interpolatedTime: number) {
            context.save();

            if (interpolatedTime < 0) interpolatedTime = 0;
            if (interpolatedTime > 1) interpolatedTime = 1;

            var _x = this.x|0;
            var _y = this.y|0;
            if (this.animations.length > 0) {
                var animationIndex = ((interpolatedTime * Constants.animationSteps) | 0);
                var animation = this.animations[animationIndex];
                if (animation) {

                    var interpolateStep = (interpolatedTime % (1 / Constants.animationSteps)) * Constants.animationSteps;
                    _x = animation.fromX + (animation.x - animation.fromX) * interpolateStep|0;
                    _y = animation.fromY + (animation.y - animation.fromY) * interpolateStep|0;
                }
            }

            var result = this.path[0];
            if (result) {

                context.lineWidth = 5;
                context.strokeStyle = "yellow";
//                context.strokeRect(result.x * Constants.squareSize, result.y * Constants.squareSize, Constants.squareSize, Constants.squareSize);

            }
            context.strokeStyle = "green";
//            context.strokeRect(this.squareX * Constants.squareSize, this.squareY * Constants.squareSize, Constants.squareSize, Constants.squareSize);



            context.lineWidth = 5;
            context.strokeStyle = "yellow";
            context.fillStyle = "red";
            context.fillRect((_x) - (Constants.squareSize / 2), (_y) - (Constants.squareSize / 2), (Constants.squareSize), (Constants.squareSize));
            context.strokeRect((_x) - (Constants.squareSize / 2), (_y) - (Constants.squareSize / 2), (Constants.squareSize ), (Constants.squareSize ));
            context.restore();
        }

        rePathFind(squareX: number, squareY: number) {
            this.rePathFindPosition = { x: squareX, y: squareY };
        }

        tick() {
            //            console.log('ticked');

            if (this.rePathFindPosition) {
                var graph = new Graph(this.game.grid);
                var start = graph.grid[this.squareX][this.squareY];
                var end = graph.grid[this.rePathFindPosition.x][this.rePathFindPosition.y];
                this.path = astar.search(graph, start, end);
                this.rePathFindPosition = undefined;
            }


            var result = this.path[0];
            this.animations = [];

            var projectedX: number;
            var projectedY: number;
            var projectedSquareX: number;
            var projectedSquareY: number;

            projectedSquareX = !result ? this.squareX : (result.x);
            projectedSquareY = !result ? this.squareY : (result.y);




            for (var i = 0; i < Constants.animationSteps; i++) {



                this.squareX = ((this.x) / Constants.squareSize) | 0;
                this.squareY = ((this.y) / Constants.squareSize) | 0;
                var fromX = this.x;
                var fromY = this.y;


                if (result && (this.squareX == result.x && this.squareY == result.y)) {
                    this.path.splice(0, 1);
                    result = this.path[0];

                    projectedSquareX = !result ? this.squareX : (result.x);
                    projectedSquareY = !result ? this.squareY : (result.y);
                }



                projectedX = projectedSquareX * Constants.squareSize + Constants.squareSize / 2;
                projectedY = projectedSquareY * Constants.squareSize + Constants.squareSize / 2;


                if ((projectedX | 0) == (this.x | 0) && (projectedY | 0) == (this.y | 0)) {
                    return;
                }

                this.x = Constants.moveTowards(this.x, projectedX, (this.speed / Constants.animationSteps));
                this.y = Constants.moveTowards(this.y, projectedY, (this.speed / Constants.animationSteps));


                this.animations.push({
                    fromX: fromX,
                    fromY: fromY,

                    x: this.x,
                    y: this.y
                });
            }


        }

    }

}

File: Pather.Client/Person.cs
using System.Collections.Generic;
using System.Html.Media.Graphics;
using Pather.Client.Definitions;
using Pather.Client.Utils;

namespace Pather.Client
{
    public class Person
    {
        public double X { get; set; }
        public double Y { get; set; }
        public int SquareX { get; set; }
        public int SquareY { get; set; }
        public double Speed { get; set; }
        public List<AStarPath> Path { get; set; }
        public Point RePathFindPosition { get; set; }
        public List<AnimationPoint> Animations { get; set; }
        private Game Game { get; set; }

        public Person(Game game)
        {
            Game = game;
            X = 0;
            Y = 0;
            SquareX = 0;
            SquareY = 0;
            Speed = 50;
            Path = new List<AStarPath>();
            RePathFindPosition = null;

        }

        public void Init(int squareX, int squareY)
        {
            SquareX = squareX;
            SquareY = squareY;
            X = SquareX * Constants.SquareSize;
            Y = SquareY * Constants.SquareSize;
            Animations = new List<AnimationPoint>();
        }

        public void RePathFind(int SquareX, int SquareY)
        {
            RePathFindPosition = new Point(SquareX, SquareY);
        }

        public void Draw(CanvasRenderingContext2D context, double interpolatedTime)
        {
            context.Save();

            if (interpolatedTime < 0) interpolatedTime = 0;
            if (interpolatedTime > 1) interpolatedTime = 1;

            var _x = (int)X;
            var _y = (int)Y;
            if (Animations.Count > 0)
            {
                var animationIndex = ((int)(interpolatedTime * Constants.AnimationSteps));
                var animation = Animations[animationIndex];
                if (animation != null)
                {

                    var interpolateStep = (interpolatedTime % (1.0 / Constants.AnimationSteps)) * Constants.AnimationSteps;
                    _x = (int)(animation.FromX + (animation.X - animation.FromX) * interpolateStep);
                    _y = (int)(animation.FromY + (animation.Y - animation.FromY) * interpolateStep);
                }
            }

            var result = Path[0];
            if (result != null)
            {

                context.LineWidth = 5;
                context.StrokeStyle = "yellow";
                //                context.StrokeRect(result.X * Constants.SquareSize, result.Y * Constants.SquareSize, Constants.SquareSize, Constants.SquareSize);

            }
            context.StrokeStyle = "green";
            //            context.StrokeRect(SquareX * Constants.SquareSize, SquareY * Constants.SquareSize, Constants.SquareSize, Constants.SquareSize);

//            Console.WriteLine(_x + " " + _y);


            context.LineWidth = 5;
            context.StrokeStyle = "yellow";
            context.FillStyle = "red";
            context.FillRect((_x) - (Constants.SquareSize / 2), (_y) - (Constants.SquareSize / 2), (Constants.SquareSize), (Constants.SquareSize));
            context.StrokeRect((_x) - (Constants.SquareSize / 2), (_y) - (Constants.SquareSize / 2), (Constants.SquareSize), (Constants.SquareSize));
            context.Restore();
        }

        public void Tick()
        {
            //            console.log('ticked');

            if (RePathFindPosition != null)
            {
                var graph = new AStarGraph(Game.Grid);
                var start = graph.Grid[SquareX][SquareY];
                var end = graph.Grid[(int)RePathFindPosition.X][(int)RePathFindPosition.Y];
                Path = new List<AStarPath>(AStar.Search(graph, start, end));
                RePathFindPosition = null;
            }


            var result = Path[0];
            Animations = new List<AnimationPoint>();

            int projectedX;
            int projectedY;
            int projectedSquareX;
            int projectedSquareY;

            projectedSquareX = result == null ? SquareX : (result.X);
            projectedSquareY = result == null ? SquareY : (result.Y);




            for (var i = 0; i < Constants.AnimationSteps; i++)
            {



                SquareX = (int)((X) / Constants.SquareSize);
                SquareY = (int)((Y) / Constants.SquareSize);
                var fromX = X;
                var fromY = Y;


                if (result != null && (SquareX == result.X && SquareY == result.Y))
                {
                    Path.RemoveAt(0);
                    result = Path[0];

                    projectedSquareX = result == null ? SquareX : (result.X);
                    projectedSquareY = result == null ? SquareY : (result.Y);
                }



                projectedX = projectedSquareX * Constants.SquareSize + Constants.SquareSize / 2;
                projectedY = projectedSquareY * Constants.SquareSize + Constants.SquareSize / 2;


                if (((int)projectedX) == ((int)X) && ((int)projectedY) == ((int)Y))
                {
                    return;
                }

                X = Constants.MoveTowards(X, projectedX, (Speed / Constants.AnimationSteps));
                Y = Constants.MoveTowards(Y, projectedY, (Speed / Constants.AnimationSteps));


                Animations.Add(new AnimationPoint(fromX, fromY, X, Y));
            }


        }

    }
}